### PR TITLE
Support model cores via option --produce-model-cores.

### LINKED ISSUE #1233: Minimal model cores
It would be interesting to have the feature that prints only the "core" of a model, that is, a subset of the model values that can be extended arbitrarily, here is an example:

(declare-fun x () Int)
(declare-fun y () Int)
(declare-fun z () Int)
(declare-fun f (Int) Int)
(assert (= (f x) 0))
(assert (or (> z 5) (> y 5)))
(check-sat)

By default we may get a model looking like this:

(model
(declare-fun x () Int 2)
(declare-fun y () Int 0)
(declare-fun z () Int 6)
(declare-fun f ((_x Int)) Int 0)
)

This could be minimized to:

(model
(declare-fun z () Int 6)
(declare-fun f ((_x Int)) Int 0)
)

Notice that

(and
(= ((\ _x 0) x) 0)
(or (> 6 5) (> y 5))
)

rewrites to 

true

(for any x,y), which our rewriter would detect. Hence the functionality of check-model + rewriter could be used to greedily minimize various for which the input still rewrites to true.

### PR DESCRIPTION
**Pull request by ajreynol**
This adds support for model cores, fixes #1233.

It includes some minor cleanup and additions to utility functions.

@barrettcw , you may be interested in this PR.

### CODE REVIEW

File: src/theory/subs_minimize.cpp [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+          Node vop = it->second;
+          if (vop.getKind() == LAMBDA)
+          {
+            visit.push_back(op);
+            Node curr = vop[0];
+            for (unsigned i = 0, size = cur.getNumChildren(); i < size; i++)
+            {
+              it = value.find(cur[i]);
+              Assert(it != value.end());
+              Node scurr = curr.substitute(cur[i], it->second);
```
Comment (4tXJ7f) [inline src/theory/subs_minimize.cpp]: Shouldn't this be `Node scurr = vop[1].substitute(curr[i], it->second);`?

File: src/theory/subs_minimize.cpp [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        bool alreadyJustified = false;
+
+        // if the operator is an apply uf, check its value
+        if (cur.getKind() == APPLY_UF)
+        {
+          Node op = cur.getOperator();
+          it = value.find(op);
+          Assert(it != value.end());
+          Node vop = it->second;
+          if (vop.getKind() == LAMBDA)
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/subs_minimize.cpp]: No, but this isn't critical for this code.

File: src/theory/subs_minimize.cpp [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+          {
+            visit.push_back(op);
+            Node curr = vop[0];
+            for (unsigned i = 0, size = cur.getNumChildren(); i < size; i++)
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/subs_minimize.cpp]: There was a bug in the code, I'll fix.  The fixed code does require an index-based for loop.

File: src/theory/subs_minimize.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/subs_minimize.cpp]: Yes, I'll fix.

File: src/theory/subs_minimize.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+   *
+   * If n { vars -> subs } rewrites to target, this method returns true, and
+   * vars[i1]...vars[in] are added to rewVars, such that
+   * n { vars[i_1] -> subs[i_1] ... vars[i_n]->subs[i_n] } also rewrites to
+   * target.
+   *
+   * If n { vars -> subs } does not rewrite to target, this method returns
+   * false.
+   */
+  bool find(Node n,
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/subs_minimize.h]: I had thought about making it have state, perhaps its not necessary for now, I'll change to static.

File: src/theory/subs_minimize.cpp [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+                    << " : " << reqVars << std::endl;
+
+  return true;
+}
+
+bool SubstitutionMinimize::isSingularArg(Node n, Kind k, unsigned arg)
+{
+  // Notice that this function is hardcoded. We could compute this function
+  // in a theory-independent way using partial evaluation. However, we
+  // prefer performance to generality here.
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/subs_minimize.cpp]: There are a couple of reasons: that function is cached for efficiency. Also, it is overkill to construct the return value here (we just need to know yes/no that it is a singular arg).

File: test/regress/regress0/model-core.smt2 [hunk 6]
```diff
@@ -0,0 +1,10 @@
+; COMMAND-LINE: --produce-model-cores
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply test/regress/regress0/model-core.smt2]: Let's do this in a separate PR if necessary.

File: src/smt/model_core_builder.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of utility for building model cores
+ **/
+
+#include "smt/model_core_builder.h"
+
+#include "theory/subs_minimize.h"
+
+using namespace CVC4::kind;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/smt/model_core_builder.cpp]: I removed kind:: from 2 places, now its used :)

File: src/smt/model_core_builder.h [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+   * It returns true if m is a model for assertions. In this case, we set:
+   *   m->usingModelCore();
+   *   m->recordModelCoreSymbol(s1); ... m->recordModelCoreSymbol(sn);
+   * such that each formula in assertions under the substitution
+   * { s1 -> m(s1), ..., sn -> m(sn) } rewrites to true.
+   *
+   * If m is not a model for assertions, this method returns false and m is
+   * left unchanged.
+   */
+  bool setModelCore(const std::vector<Expr>& assertions, Model* m);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/smt/model_core_builder.h]: Also updating this to static.

File: src/theory/subs_minimize.cpp [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+
+namespace CVC4 {
+namespace theory {
+
+SubstitutionMinimize::SubstitutionMinimize() {}
+
+bool SubstitutionMinimize::find(Node n,
+                                Node target,
+                                const std::vector<Node>& vars,
+                                const std::vector<Node>& subs,
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/subs_minimize.cpp]: We could support both, but I'd prefer keeping this as vectors.  If we used this class to do many minimizations of explanations of small substitutions, then vectors are actually way more efficient than unordered_map.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/subs_minimize.cpp]: Notice that this is related to which Node::substitute(...) method should be default. There was a recent attempt by Tim to make an unordered_map version default, but this was noticeably slower for quantifiers than the vector versions.  https://github.com/CVC4/CVC4/pull/2304

File: src/theory/subs_minimize.cpp [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+    visit.pop_back();
+    it = value.find(cur);
+
+    if (it == value.end())
+    {
+      if (cur.isVar())
+      {
+        const std::vector<Node>::const_iterator& it =
+            std::find(vars.begin(), vars.end(), cur);
+        if (it == vars.end())
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/subs_minimize.cpp]: No, this class is more general than just for use in the model core code.  We could be reasoning about partial substitutions here.  For example, we could ask:

Minimize { y -> 0, z -> 0 }
such that y+x rewrites to x.

Where x is not given a value.

**Comment by ajreynol** [comment]
Bump up to major, since UTRC needs this.

File: src/smt/model_core_builder.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/smt/model_core_builder.cpp]: That's one way to solve it :)

File: src/theory/subs_minimize.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/subs_minimize.cpp]: Great, thanks!

File: test/regress/regress0/model-core.smt2 (hunk 6 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply test/regress/regress0/model-core.smt2]: Sounds good, yes.

File: src/theory/subs_minimize.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/subs_minimize.cpp]: Ok, we can leave it for now.

File: src/theory/subs_minimize.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/subs_minimize.cpp]: Yeah, I see that now.

File: src/theory/subs_minimize.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/subs_minimize.cpp]: That makes sense.

**Pull request merged**